core/dbt/task/test.py

```diff
@@ -78,9 +67,47 @@ def execute_schema_test(self, test: CompiledSchemaTestNode):
     def before_execute(self):
         self.print_start_line()
 
+    def execute_data_test(
+        self,
+        test: CompiledDataTestNode,
+        manifest: Manifest
+    ) -> int:
+        context = generate_runtime_model(
+            test, self.config, manifest
+        )
+
+        materialization_macro = manifest.find_materialization_macro_by_name(
+            self.config.project_name,
+            test.get_materialization(),
+            self.adapter.type())
+
+        if materialization_macro is None:
+            missing_materialization(test, self.adapter.type())
+
+        if 'config' not in context:
+            raise InternalException(
+                'Invalid materialization context generated, missing config: {}'
+                .format(context)
+            )
+
+        MacroGenerator(materialization_macro, context)()
+        result = context['load_result']('main')
+        table = result['table']
+        num_rows = len(table.rows)
+        if num_rows != 1:
+            num_cols = len(table.columns)
+            # since we just wrapped our query in `select count(*)`, we are in
+            # big trouble!
+            raise InternalException(
+                f"dbt internally failed to execute {test.unique_id}: "
+                f"Returned {num_rows} rows and {num_cols} cols, but expected "
+                f"1 row and 1 column"
+            )
+        return int(table[0][0])
```

Comment:
Rather than just returning a single number to the `execute` method, I imagine the materialization-calling method should construct and return the full `RunResult`, similar to [`_build_run_model_result`](https://github.com/fishtown-analytics/dbt/blob/440a5e49e274b7c98a726c1acdd202fbf1063df1/core/dbt/task/run.py#L192). It seems like that will enable us to manage the dynamic `message` interpolation, as well as recording the `adapter_response`.

Of course, those can and should be new issues! It's cool to see the path taking shape.